### ISSUE TITLE
ReconnectingClientFactory should wrap the protocol maker with a connectionMade that calls resetDelay for you

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1994]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#1994|
|Type|enhancement|
|Created|2006-08-13 01:31:19Z|

Why rely on a protocol to call resetDelay? I propose something like this (the implementation is sloppy, but you get the idea):

```
class SmartReconnectingClientFactory(ReconnectingClientFactory):

    def buildProtocol(self, *args, **kw):
        prot = ReconnectingClientFactory.buildProtocol(self, *args, **kw)
        old_connectionMade = prot.connectionMade
        def connectionMade(*a, **kw):
            self.resetDelay()
            return old_connectionMade(*a, **kw)
        prot.connectionMade = connectionMade
        return prot    
```

<details><summary>Searchable metadata</summary>

```
trac-id__1994 1994
type__enhancement enhancement
reporter__ghazel ghazel
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__core core
keywords__ 
time__1155432679000000 1155432679000000
changetime__1265386476000000 1265386476000000
version__None None
owner__ghazel ghazel

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Perhaps ReconnectingClientFactory should be replaced with a WrappingFactory of some sort?

**Comment by twisted-trac** [comment]
|<img alt="naked's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|naked commented|
|-|-|

I assumed the resetDelay() call was supposed to be made only after all the connection setup work has been done, not simply because a TCP connection was established. Otherwise this would hammer the server if the server replies "not now" just after connecting.

A useful default would be that connectionMade causes resetDelay, but should be overridable, shouldn't it?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @ghazel|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

naked is right.  Sometimes you have a server that accepts a connection and then very quickly drops it for whatever reason.  It needs to be possible for `resetDelay` to be called only later, after some arbitrary point where the protocol has decided things have settled down.
